gflownet/evaluator/base.py

```diff
@@ -0,0 +1,793 @@
+import copy
+import os
+import pickle
+import time
+from collections import defaultdict
+from typing import Union
+
+import numpy as np
+import torch
+from scipy.special import logsumexp
+
+from gflownet.utils.batch import Batch
+from gflownet.utils.common import (
+    batch_with_rest,
+    load_gflow_net_from_run_path,
+    tfloat,
+    torch2np,
+)
+
+_sentinel = object()
```

Comment:
What is the purpose or role of this?

Reply:
Typically, a "sentinel" is used to allow arbitrary user inputs.

The use-case is: if you have a function that can be used with any input type, how do you know if the user has provided an argument or not?

Example:

```
def f(a=None):
    print("User wants:", a)

f(a)
f()
```

Reply:
I see! Thanks for explaining that!

Reply:
(and leading `_` to signal it's private and should not be used by a user of this library. The only place outside the code-base it can be --and is-- used is in tests)